modules/dlsobackend/dlsobackend.cc

```diff
@@ -0,0 +1,297 @@
+  this->api.get_metas = (bool (*)(void * handle, uint8_t qlen, const char * qname, fill_metas_cb_t cb, void * metas)) dlsym(this->dlhandle, "pdns_dlso_domain_all_metadatas");
+  this->api.get_meta = (bool (*)(void * handle, uint8_t qlen, const char * qname, uint8_t meta_len, const char * kind, fill_meta_cb_t cb, void * meta)) dlsym(this->dlhandle, "pdns_dlso_domain_metadata");
+
+  this->api.get_before_after = (bool (*)(void * handle, uint32_t domain_id, uint8_t qlen, const char * qname, fill_before_after_t cb, void * beforeAfter)) dlsym(this->dlhandle, "pdns_dlso_get_before_after");
+  this->api.get_tsig_key = (bool (*)(void * handle, uint8_t qlen, const char * qname, uint8_t alg_len, const char * alg, fill_tsig_key_cb_t cb, void * content)) dlsym(this->dlhandle, "pdns_dlso_get_tsig_key");
+
+  this->handle = this->api.new_(this->d_dnssec, args.c_str());
+
+  if (this->handle == NULL)
+    throw PDNSException("Failed to initialize dlso");
+}
+
+DlsoBackend::~DlsoBackend() {
+  this->api.free_(this->handle);
+  dlclose(this->dlhandle);
+}
+
+/**
+ * The functions here are just remote json stubs that send and receive the method call
+ * data is mainly left alone, some defaults are assumed.
+ */
+void DlsoBackend::lookup(const QType &qtype, const DNSName& qdomain, DNSPacket *pkt_p, int zoneId) {
+  if(in_query)
+    throw PDNSException("Attempt to lookup while one running");
+
+  string qname = qdomain.toString();
+  bool success;
+
+  if (pkt_p != NULL) {
+    ComboAddress edns_or_resolver_ip = pkt_p->getRealRemote().getNetwork();
+    success = api.lookup(handle, qtype.getCode(), qname.size(), qname.c_str(), (sockaddr*) &edns_or_resolver_ip.sin4);
+  } else {
+    success = api.lookup(handle, qtype.getCode(), qname.size(), qname.c_str(), NULL);
+  }
+
+  if (!success)
+    throw PDNSException("Backend failed");
+}
+
+bool DlsoBackend::list(const DNSName& target, int domain_id, bool include_disabled) {
+
+  string qname = target.toString();
+  bool success;
+
+  success = api.list(handle, qname.size(), qname.c_str());
+
+  if (!success)
+    throw PDNSException("Backend failed");
+
+  return success;
+}
+
+void fill_cb(DNSResourceRecord *rr, const struct resource_record *record) {
+  rr->qtype = record->qtype;
+  rr->qname = DNSName(string(record->qname, record->qname_len));
+  rr->qclass = QClass::IN;
+  rr->content = string(record->content, record->content_len);
+  rr->ttl = record->ttl;
+  rr->auth = record->auth;
+  rr->scopeMask = record->scopeMask;
+  rr->domain_id = record->domain_id;
+}
+
+bool DlsoBackend::get(DNSResourceRecord &rr) {
+  bool success = api.get(handle, fill_cb, &rr);
+
+  if (!success) {
+    in_query = false;
+    return false;
+  }
+
+  return true;
+}
+
+void fill_meta_cb(std::vector<std::string>* meta, uint8_t value_len, const struct dns_meta_value * values) {
+  for (int i=0; i<value_len; i++) {
```

Comment:
The loop index should be unsigned, otherwise the compiler will complain about comparing signed and unsigned values.